File: android/app/src/main/java/org/kog/player/KogApi.kt
package org.kog.player

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

data class Track(
    val kind: String,
    val path: String,
    val entry: String = "",
    val fragment: String = "",
    val name: String = "",
    val title: String = "",
    val artist: String = "",
    val album: String = "",
    val duration: Long = 0,
) {
    val key: String get() = "$kind|$path|$entry|$fragment"
    val label: String get() = title.ifBlank {
        name.ifBlank { entry.ifBlank { path }.substringAfterLast('/') }
    }
    val detail: String get() = listOf(artist, album).filter(String::isNotBlank).joinToString(" · ")
    val isDevice: Boolean get() = kind == "device"

    fun locator(): JSONObject = JSONObject()
        .put("kind", kind).put("path", path).put("entry", entry).put("fragment", fragment)

    fun saved(): JSONObject = locator().put("name", name).put("title", title)
        .put("artist", artist).put("album", album).put("duration", duration)

    companion object {
        fun parse(row: JSONObject): Track = Track(
            kind = row.optString("kind", "local"),
            path = row.optString("path"),
            entry = row.optString("entry", ""),
            fragment = if (row.isNull("fragment")) "" else row.optString("fragment", ""),
            name = row.optString("name", ""),
            title = row.optString("title", ""),
            artist = row.optString("artist", ""),
            album = row.optString("album", ""),
            duration = row.optLong("duration", 0),
        )
    }
}

data class Folder(val name: String, val path: String)
data class Listing(val path: String, val parent: String, val directories: List<Folder>, val files: List<Track>)
data class SavedPlaylist(val id: Long, val name: String, val count: Int)
data class SearchPage(val tracks: List<Track>, val folders: List<Folder>, val generation: Long,
                      val count: Int, val scanned: Int, val done: Boolean)

/** The mobile client uses the same HTTP endpoints and locator shape as Kog Web. */
class KogApi(private val context: Context) {
    private val prefs = context.getSharedPreferences("kog", Context.MODE_PRIVATE)
    var server: String
        get() = prefs.getString("server", "") ?: ""
        set(value) {
            val address = value.trim().trimEnd('/')
            prefs.edit().putString("server", if (address.isNotBlank() && "://" !in address)
                "http://$address" else address).apply()
        }
    var token: String
        get() = prefs.getString("token", "") ?: ""
        set(value) { prefs.edit().putString("token", value.trim()).apply() }
    var username: String
        get() = prefs.getString("username", "") ?: ""
        set(value) { prefs.edit().putString("username", value).apply() }
    var password: String
        get() = prefs.getString("password", "") ?: ""
        set(value) { prefs.edit().putString("password", value).apply() }
    var codec: String
        get() = prefs.getString("codec", "aac") ?: "aac"
        set(value) { prefs.edit().putString("codec", value).apply() }

    fun uri(endpoint: String, vararg params: Pair<String, String>): String {
        val origin = server.ifBlank { "http://127.0.0.1:8420" }
        val builder = Uri.parse("$origin$endpoint").buildUpon()
        for ((key, value) in params) builder.appendQueryParameter(key, value)
        return builder.build().toString()
    }

    fun stream(track: Track): String = if (track.isDevice) track.path else uri(
        "/api/stream", "kind" to track.kind, "path" to track.path,
        "entry" to track.entry, "fragment" to track.fragment, "codec" to codec,
        "token" to token,
    )

    fun art(track: Track): String? = if (track.isDevice) null else uri(
        "/api/art", "kind" to track.kind, "path" to track.path, "token" to token,
    )

    private suspend fun request(path: String, method: String = "GET", body: Any? = null): String =
        withContext(Dispatchers.IO) {
            val connection = URL(path).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = method
                connection.connectTimeout = 10_000
                connection.readTimeout = 45_000
                val authorization = when {
                    token.isNotBlank() -> "Bearer $token"
                    username.isNotBlank() -> "Basic " + Base64.getEncoder().encodeToString(
                        "$username:$password".toByteArray(Charsets.UTF_8))
                    else -> ""
                }
                if (authorization.isNotEmpty()) connection.setRequestProperty("Authorization", authorization)
                if (body != null) {
                    connection.doOutput = true
                    connection.setRequestProperty("Content-Type", "application/json")
                    connection.outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
                }
                val response = if (connection.responseCode in 200..299) connection.inputStream else connection.errorStream
                val text = response?.bufferedReader()?.use { it.readText() }.orEmpty()
                if (connection.responseCode !in 200..299) {
                    val message = runCatching { JSONObject(text).optString("error") }.getOrNull()
                    error(message?.ifBlank { null } ?: "HTTP ${connection.responseCode}")
                }
                text
            } finally {
                connection.disconnect()
            }
        }

    suspend fun health(): Boolean = runCatching { request(uri("/api/health")); true }.getOrDefault(false)

    suspend fun browse(path: String = ""): Listing {
        val response = JSONObject(request(uri("/api/library", "path" to path)))
        return Listing(
            response.optString("path"), response.optString("parent", ""),
            response.optJSONArray("directories").objects().map {
                Folder(it.optString("name"), it.optString("path"))
            },
            response.optJSONArray("files").objects().map(Track::parse),
        )
    }

    suspend fun collect(path: String): List<Track> {
        val tracks = JSONObject(request(uri("/api/library/collect", "path" to path)))
            .optJSONArray("tracks").objects().map(Track::parse)
        return withMetadata(tracks)
    }

    suspend fun expand(track: Track): List<Track> {
        val response = JSONObject(request(uri("/api/expand"), "POST", JSONArray().put(
            track.locator().put("name", track.name))))
        val rows = response.optJSONArray("tracks")?.optJSONArray(0)
        return withMetadata(rows.objects().map(Track::parse))
    }

    suspend fun withMetadata(tracks: List<Track>): List<Track> {
        if (tracks.isEmpty()) return tracks
        return tracks.chunked(100).flatMap { chunk ->
            val body = JSONArray()
            chunk.forEach { body.put(it.locator()) }
            val rows = JSONArray(request(uri("/api/metadata"), "POST", body))
            chunk.mapIndexed { index, track ->
                val row = rows.optJSONObject(index)
                track.copy(
                    title = row?.optString("title", "")?.takeUnless { it == "null" }.orEmpty(),
                    artist = row?.optString("artist", "")?.takeUnless { it == "null" }.orEmpty(),
                    album = row?.optString("album", "")?.takeUnless { it == "null" }.orEmpty(),
                    duration = ((row?.optDouble("duration", 0.0) ?: 0.0) * 1000).toLong(),
                )
            }
        }
    }

    suspend fun search(query: String): SearchPage = searchPage("/api/library/search", "q" to query)
    suspend fun more(generation: Long, offset: Int): SearchPage = searchPage(
        "/api/library/search/more", "g" to generation.toString(), "offset" to offset.toString())

    private suspend fun searchPage(path: String, vararg params: Pair<String, String>): SearchPage {
        val response = JSONObject(request(uri(path, *params)))
        val rows = response.optJSONArray("results").objects()
        return SearchPage(
            rows.filterNot { it.optBoolean("is_dir") }.map(Track::parse),
            rows.filter { it.optBoolean("is_dir") }.map {
                Folder(it.optString("name"), it.optString("path"))
            }, response.optLong("generation"), response.optInt("total"),
            response.optInt("scanned"), response.optBoolean("done"),
        )
    }

    suspend fun playlists(): List<SavedPlaylist> = JSONObject(request(uri("/api/playlists")))
        .optJSONArray("playlists").objects().map {
            SavedPlaylist(it.optLong("id"), it.optString("name"), it.optInt("entryCount"))
        }

    suspend fun playlist(id: Long): List<Track> {
        val tracks = JSONObject(request(uri("/api/playlists/$id")))
            .optJSONArray("entries").objects().map(Track::parse)
        return withMetadata(tracks)
    }

    suspend fun createPlaylist(name: String) { request(uri("/api/playlists"), "POST", JSONObject().put("name", name)) }
    suspend fun renamePlaylist(id: Long, name: String) {
        request(uri("/api/playlists/$id/rename"), "POST", JSONObject().put("name", name))
    }
    suspend fun deletePlaylist(id: Long) { request(uri("/api/playlists/$id"), "DELETE") }
    suspend fun appendPlaylist(id: Long, tracks: List<Track>) {
        val entries = JSONArray()
        tracks.filterNot(Track::isDevice).forEach { entries.put(it.locator()) }
        if (entries.length() > 0) request(uri("/api/playlists/$id/entries"), "POST",
            JSONObject().put("entries", entries))
    }

    suspend fun stars(): Set<String> = JSONObject(request(uri("/api/stars")))
        .optJSONArray("entries").objects().map { Track.parse(it).key }.toSet()

    suspend fun star(track: Track, starred: Boolean) {
        request(uri("/api/stars"), "POST", track.locator().put("starred", starred))
    }

    suspend fun radio(enabled: Boolean, root: String): List<Track> {
        val response = JSONObject(request(uri("/api/radio/enabled", "root" to root), "POST",
            JSONObject().put("enabled", enabled)))
        return if (enabled) withMetadata(response.optJSONArray("entries").objects().map(Track::parse)) else emptyList()
    }

    suspend fun radioAdvance(root: String): List<Track> = withMetadata(
        JSONObject(request(uri("/api/radio/advance", "root" to root), "POST"))
            .optJSONArray("entries").objects().map(Track::parse))
}

private fun JSONArray?.objects(): List<JSONObject> = if (this == null) emptyList() else
    (0 until length()).mapNotNull(::optJSONObject)

File: ios/Kog/KogAPI.swift
import Foundation
import Security

enum Secrets {
    static func read(_ name: String) -> String {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
                                    kSecAttrService as String: "org.kog.player",
                                    kSecAttrAccount as String: name,
                                    kSecReturnData as String: true,
                                    kSecMatchLimit as String: kSecMatchLimitOne]
        var result: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
              let data = result as? Data else { return "" }
        return String(data: data, encoding: .utf8) ?? ""
    }

    static func write(_ name: String, _ value: String) {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
                                    kSecAttrService as String: "org.kog.player",
                                    kSecAttrAccount as String: name]
        SecItemDelete(query as CFDictionary)
        guard !value.isEmpty else { return }
        var attributes = query
        attributes[kSecValueData as String] = Data(value.utf8)
        SecItemAdd(attributes as CFDictionary, nil)
    }
}

struct KogAPI {
    var server: String
    var token: String
    var username: String
    var password: String
    var codec: String

    func url(_ endpoint: String, _ query: [String: String] = [:]) throws -> URL {
        let origin = server.contains("://") ? server : "http://\(server)"
        guard var components = URLComponents(string: origin.trimmingCharacters(in: .whitespacesAndNewlines)),
              components.host != nil else { throw KogError.invalidServer }
        let prefix = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = prefix.isEmpty ? endpoint : "/\(prefix)\(endpoint)"
        components.queryItems = query.isEmpty ? nil : query.sorted { $0.key < $1.key }.map { URLQueryItem(name: $0.key, value: $0.value) }
        guard let url = components.url else { throw KogError.invalidServer }
        return url
    }

    func stream(_ track: Track) throws -> URL {
        if track.isDevice { return URL(fileURLWithPath: track.path) }
        return try url("/api/stream", track.locator.merging(["codec": codec, "token": token]) { _, new in new })
    }

    func art(_ track: Track) -> URL? {
        guard !track.isDevice else { return nil }
        return try? url("/api/art", ["kind": track.kind, "path": track.path, "token": token])
    }

    func artData(_ track: Track) async throws -> Data {
        try await request("/api/art", query: ["kind": track.kind, "path": track.path])
    }

    private func authenticatedRequest(_ endpoint: String, query: [String: String] = [:]) throws -> URLRequest {
        var request = URLRequest(url: try url(endpoint, query))
        request.timeoutInterval = 45
        if !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else if !username.isEmpty {
            let raw = Data("\(username):\(password)".utf8).base64EncodedString()
            request.setValue("Basic \(raw)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    func download(_ track: Track) async throws -> (URL, String) {
        guard track.kind == "local" || track.kind == "archive" else {
            throw KogError.response("Only server files and archive members can be saved on this iPhone")
        }
        var request = try authenticatedRequest("/api/media/download", query: [
            "kind": track.kind, "path": track.path, "entry": track.entry,
        ])
        request.timeoutInterval = 3600
        let (temporary, response) = try await URLSession.shared.download(for: request)
        guard let response = response as? HTTPURLResponse else {
            throw KogError.response("No server response")
        }
        guard (200..<300).contains(response.statusCode) else {
            throw KogError.response("Download failed (HTTP \(response.statusCode))")
        }
        let sourceName = track.kind == "archive" ? track.entry : track.path
        let filename = URL(fileURLWithPath: sourceName).lastPathComponent
        guard !filename.isEmpty && filename != "." && filename != ".." else {
            throw KogError.response("The server file has no usable name")
        }
        return (temporary, filename)
    }

    private func request(_ endpoint: String, query: [String: String] = [:],
                         method: String = "GET", body: Any? = nil) async throws -> Data {
        var request = try authenticatedRequest(endpoint, query: query)
        request.httpMethod = method
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let response = response as? HTTPURLResponse else { throw KogError.response("No server response") }
        guard (200..<300).contains(response.statusCode) else {
            let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
            throw KogError.response((object?["error"] as? String) ?? "HTTP \(response.statusCode)")
        }
        return data
    }

    func health() async throws { _ = try await request("/api/health") }
    func browse(_ path: String = "") async throws -> Listing {
        try JSONDecoder().decode(Listing.self, from: await request("/api/library", query: ["path": path]))
    }
    func collect(_ path: String) async throws -> [Track] {
        let data = try await request("/api/library/collect", query: ["path": path])
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["tracks"] ?? []
        return try await metadata(decodeTracks(rows))
    }
    func expand(_ track: Track) async throws -> [Track] {
        let data = try await request("/api/expand", method: "POST", body: [track.locator.merging(["name": track.name]) { _, new in new }])
        let tracks = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["tracks"] as? [Any]
        return try await metadata(decodeTracks(tracks?.first ?? []))
    }
    func metadata(_ tracks: [Track]) async throws -> [Track] {
        guard !tracks.isEmpty else { return [] }
        var result = [Track]()
        for start in stride(from: 0, to: tracks.count, by: 100) {
            let chunk = Array(tracks[start..<min(start + 100, tracks.count)])
            let data = try await request("/api/metadata", method: "POST", body: chunk.map(\.locator))
            let rows = (try JSONSerialization.jsonObject(with: data) as? [[String: Any]]) ?? []
            for (index, var track) in chunk.enumerated() {
                if index < rows.count {
                    let row = rows[index]
                    track.title = row["title"] as? String ?? ""
                    track.artist = row["artist"] as? String ?? ""
                    track.album = row["album"] as? String ?? ""
                    track.duration = Int64(((row["duration"] as? NSNumber)?.doubleValue ?? 0) * 1000)
                }
                result.append(track)
            }
        }
        return result
    }

    func search(_ term: String) async throws -> SearchPage {
        try JSONDecoder().decode(SearchPage.self, from: await request("/api/library/search", query: ["q": term]))
    }
    func more(_ generation: Int64, offset: Int) async throws -> SearchPage {
        try JSONDecoder().decode(SearchPage.self, from: await request("/api/library/search/more", query: ["g": "\(generation)", "offset": "\(offset)"]))
    }
    func playlists() async throws -> [SavedPlaylist] {
        let data = try await request("/api/playlists")
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["playlists"] ?? []
        let decodedData = try JSONSerialization.data(withJSONObject: rows)
        return try decodedData.withDecoded([SavedPlaylist].self)
    }
    func playlist(_ id: Int64) async throws -> [Track] {
        let data = try await request("/api/playlists/\(id)")
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["entries"] ?? []
        return try await metadata(decodeTracks(rows))
    }
    func createPlaylist(_ name: String) async throws { _ = try await request("/api/playlists", method: "POST", body: ["name": name]) }
    func renamePlaylist(_ id: Int64, name: String) async throws { _ = try await request("/api/playlists/\(id)/rename", method: "POST", body: ["name": name]) }
    func deletePlaylist(_ id: Int64) async throws { _ = try await request("/api/playlists/\(id)", method: "DELETE") }
    func appendPlaylist(_ id: Int64, tracks: [Track]) async throws {
        let entries = tracks.filter { !$0.isDevice }.map(\.locator)
        if !entries.isEmpty { _ = try await request("/api/playlists/\(id)/entries", method: "POST", body: ["entries": entries]) }
    }
    func stars() async throws -> Set<String> {
        let data = try await request("/api/stars")
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["entries"] ?? []
        return Set(decodeTracks(rows).map(\.id))
    }
    func star(_ track: Track, enabled: Bool) async throws {
        var body: [String: Any] = track.locator
        body["starred"] = enabled
        _ = try await request("/api/stars", method: "POST", body: body)
    }
    func radio(_ enabled: Bool, root: String) async throws -> [Track] {
        let data = try await request("/api/radio/enabled", query: ["root": root], method: "POST", body: ["enabled": enabled])
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["entries"] ?? []
        return enabled ? try await metadata(decodeTracks(rows)) : []
    }
    func radioAdvance(root: String) async throws -> [Track] {
        let data = try await request("/api/radio/advance", query: ["root": root], method: "POST")
        let rows = (try JSONSerialization.jsonObject(with: data) as? [String: Any])?["entries"] ?? []
        return try await metadata(decodeTracks(rows))
    }

    private func decodeTracks(_ object: Any) -> [Track] {
        guard let rows = object as? [[String: Any]],
              let data = try? JSONSerialization.data(withJSONObject: rows) else { return [] }
        return (try? JSONDecoder().decode([Track].self, from: data)) ?? []
    }
}

private extension Data {
    func withDecoded<T: Decodable>(_ type: T.Type) throws -> T { try JSONDecoder().decode(type, from: self) }
}
